### ISSUE TITLE
`Ed25519KeyPair::generate_pkcs8` generates malformed PKCS#8 v2 data.

### ISSUE THREAD

**Issue by ShadowJonathan** [briansmith/ring#1299]
(Found [while adding PKCS#8 v2 support on `pkcs8`](https://github.com/RustCrypto/utils/issues/419).)

<details>
<summary>HEX data</summary>

```
30 53 02 01  01 30 05 06  03 2B 65 70  04 22 04 20  3A 13 3D AB  AD A2 AA 9C  E5 4B 09 61  CC 3F 15 76  B0 94 3D C8  6E BF 72 A5  6E 05 2C 43  F3 0F A3 A5  A1 23 03 21  00 A3 A7 EA  E3 A8 37 38  30 BC 47 E1  16 7B C5 0E  1D B5 51 99  96 51 E0 E2  DC 58 76 23  43 8E AC 3F  31
```

</details>

Putting this data into two different external ASN.1 decoders yields two different results;

- https://holtstrom.com/michael/tools/asn1decoder.php
- https://lapo.it/asn1js/#MFMCAQEwBQYDK2VwBCIEIDoTPautoqqc5UsJYcw_FXawlD3Ibr9ypW4FLEPzD6OloSMDIQCjp-rjqDc4MLxH4RZ7xQ4dtVGZllHg4txYdiNDjqw_MQ

In the first, we get a 34-byte slice of `04203A133DABADA2AA9CE54B0961CC3F1576B0943DC86EBF72A56E052C43F30FA3A5`, and in the second (besides a [buggy interface](https://user-images.githubusercontent.com/22740616/119045069-4609c380-b9bb-11eb-87a2-0ca76bed61c6.png)), we get a 32-byte slice of `3A133DABADA2AA9CE54B0961CC3F1576B0943DC86EBF72A56E052C43F30FA3A5`

It should be noted that the valid length for a private key is 32 bytes.

I've looked at the [RFC detailing how this should look like](https://datatracker.ietf.org/doc/html/rfc5958#section-2), i've looked at the definition of the `OCTET STRING`, i've used another library, `simple_asn1`, which gave the same result;

<details>
<summary>dumped data</summary>

```
Ok([Sequence(00, [Integer(02, 01), Sequence(05, [ObjectIdentifier(07, OID([01, 03, 101, 112]))]), OctetString(0C, [04, 20, 3A, 13, 3D, AB, AD, A2, AA, 9C, E5, 4B, 09, 61, CC, 3F, 15, 76, B0, 94, 3D, C8, 6E, BF, 72, A5, 6E, 05, 2C, 43, F3, 0F, A3, A5]), Explicit(ContextSpecific, 30, 01, BitString(32, 100, [A3, A7, EA, E3, A8, 37, 38, 30, BC, 47, E1, 16, 7B, C5, 0E, 1D, B5, 51, 99, 96, 51, E0, E2, DC, 58, 76, 23, 43, 8E, AC, 3F, 31]))])])
```

> `OctetString(0C, [04, 20, 3A, 13,`

</details>

This is data generated via the following MVP:

```rust
use ring::signature::Ed25519KeyPair;

fn main() {
    let bytes = Ed25519KeyPair::generate_pkcs8(&ring::rand::SystemRandom::new())
        .unwrap()
        .as_ref()
        .to_vec();

    std::fs::write("./tmp", bytes).unwrap();

    eprintln!("{:02X?}", simple_asn1::from_der(&bytes));
}
```

---

In short; `ed25519_pkcs8_v2_template.der` is malformed, `ring::ec::curve25519::ed25519::signing::PKCS8_TEMPLATE` is wrong, and nobody noticed up until now because [even `openssl` doesn't support PKCS#8 v2 yet](https://github.com/openssl/openssl/issues/10468).

All "PKCS#8 documents" generated up until this point (via this method) are malformed.

**Comment by ShadowJonathan** [comment]
It appears i was mistaken, but still, I find it incredibly confusing to stumble upon this.

For future reference, [this draft](https://datatracker.ietf.org/doc/html/draft-ietf-curdle-pkix-07#section-10.3) tells something about the proper markup of these keys, and that this behaviour is - albeit confusing - intended.
